File: sdks/java/src/dev/lunora/Json.java
package dev.lunora;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * A minimal JSON reader/writer.
 *
 * <p>Hand-rolled because Java SE ships no JSON at all, and pulling Jackson or Gson in would make
 * the transport's only dependency a parser it barely uses. The risk that usually makes hand-rolling
 * a bad idea does not apply here: the golden fixtures under {@code protocol/fixtures} are
 * themselves a demanding parser suite, so a mistake in this file fails the conformance tests rather
 * than hiding.
 *
 * <p>Values map to plain Java types, matching every other Lunora port: {@code null}, {@link
 * Boolean}, {@link Double}, {@link String}, {@code List<Object>} and {@code LinkedHashMap<String,
 * Object>} (insertion ordered, because object field order must survive a round trip).
 */
public final class Json {
    /**
     * Levels of ENVELOPE a wire value can sit under before its own nesting starts. Every payload
     * arrives wrapped: {@code {"result": V}} is one, and the two deepest — a batch response {@code
     * {"results":[{"result": V}]}} and a poke part {@code {"rowsPatch":[{"value": V}]}} — are
     * three.
     */
    private static final int MAX_ENVELOPE_DEPTH = 3;

    /**
     * The parser's own nesting cap, counted from the DOCUMENT root.
     *
     * <p>It cannot be {@link Wire#MAX_DEPTH} directly: that bounds a wire VALUE, whose root is
     * already {@link #MAX_ENVELOPE_DEPTH} levels into the frame carrying it, so charging the
     * envelope against the value's budget refused frames whose payload the reference encodes
     * happily. The cap is here to keep a hostile frame off the stack, and a few extra levels cost
     * nothing against that.
     */
    static final int MAX_DEPTH = Wire.MAX_DEPTH + MAX_ENVELOPE_DEPTH;

    private Json() {}

    public static Object parse(String text) {
        Parser parser = new Parser(text);
        Object value = parser.readValue();

        parser.skipWhitespace();

        if (!parser.atEnd()) {
            throw new IllegalArgumentException("json: trailing content at offset " + parser.offset);
        }

        return value;
    }

    public static String write(Object value) {
        StringBuilder out = new StringBuilder();

        writeValue(out, value);

        return out.toString();
    }

    private static void writeValue(StringBuilder out, Object value) {
        if (value == null) {
            out.append("null");
        } else if (value instanceof Boolean bool) {
            out.append(bool ? "true" : "false");
        } else if (value instanceof Number number) {
            out.append(Key.formatNumber(number.doubleValue()));
        } else if (value instanceof String text) {
            out.append(Key.jsonString(text));
        } else if (value instanceof List<?> items) {
            out.append('[');

            for (int index = 0; index < items.size(); index++) {
                if (index > 0) {
                    out.append(',');
                }

                writeValue(out, items.get(index));
            }

            out.append(']');
        } else if (value instanceof Map<?, ?> fields) {
            out.append('{');

            boolean first = true;

            for (Map.Entry<?, ?> entry : fields.entrySet()) {
                if (!first) {
                    out.append(',');
                }

                first = false;
                out.append(Key.jsonString(String.valueOf(entry.getKey())));
                out.append(':');
                writeValue(out, entry.getValue());
            }

            out.append('}');
        } else {
            throw new IllegalArgumentException(
                    "json: cannot write a " + value.getClass().getName());
        }
    }

    private static final class Parser {
        private final String text;
        private int offset;
        private int depth;

        Parser(String text) {
            this.text = text;
        }

        boolean atEnd() {
            return offset >= text.length();
        }

        void skipWhitespace() {
            while (offset < text.length() && Character.isWhitespace(text.charAt(offset))) {
                offset++;
            }
        }

        Object readValue() {
            skipWhitespace();

            if (atEnd()) {
                throw new IllegalArgumentException("json: unexpected end of input");
            }

            // Bounded here, not after parsing. Wire.MAX_DEPTH is applied to the
            // decoded tree and so never sees the payload; without this, a deeply
            // nested frame overflows the stack — and StackOverflowError is an
            // Error, which Client.handleFrame's `catch (RuntimeException)`
            // cannot catch, so it escapes onto the socket reader thread.
            if (depth > MAX_DEPTH) {
                throw new IllegalArgumentException(
                        "json: nesting exceeds the " + MAX_DEPTH + "-level limit");
            }

            char character = text.charAt(offset);

            return switch (character) {
                case '{' -> readObject();
                case '[' -> readArray();
                case '"' -> readString();
                case 't' -> readLiteral("true", Boolean.TRUE);
                case 'f' -> readLiteral("false", Boolean.FALSE);
                case 'n' -> readLiteral("null", null);
                default -> readNumber();
            };
        }

        private Object readLiteral(String literal, Object value) {
            if (!text.startsWith(literal, offset)) {
                throw new IllegalArgumentException("json: invalid literal at offset " + offset);
            }

            offset += literal.length();

            return value;
        }

        private Map<String, Object> readObject() {
            Map<String, Object> fields = new LinkedHashMap<>();

            depth++;
            offset++; // '{'
            skipWhitespace();

            if (!atEnd() && text.charAt(offset) == '}') {
                offset++;
                depth--;

                return fields;
            }

            while (true) {
                skipWhitespace();

                String key = readString();

                skipWhitespace();

                if (atEnd() || text.charAt(offset) != ':') {
                    throw new IllegalArgumentException("json: expected ':' at offset " + offset);
                }

                offset++;
                fields.put(key, readValue());
                skipWhitespace();

                if (atEnd()) {
                    throw new IllegalArgumentException("json: unterminated object");
                }

                char character = text.charAt(offset++);

                if (character == '}') {
                    depth--;

                    return fields;
                }

                if (character != ',') {
                    throw new IllegalArgumentException(
                            "json: expected ',' or '}' at offset " + (offset - 1));
                }
            }
        }

        private List<Object> readArray() {
            List<Object> items = new ArrayList<>();

            depth++;
            offset++; // '['
            skipWhitespace();

            if (!atEnd() && text.charAt(offset) == ']') {
                offset++;
                depth--;

                return items;
            }

            while (true) {
                items.add(readValue());
                skipWhitespace();

                if (atEnd()) {
                    throw new IllegalArgumentException("json: unterminated array");
                }

                char character = text.charAt(offset++);

                if (character == ']') {
                    depth--;

                    return items;
                }

                if (character != ',') {
                    throw new IllegalArgumentException(
                            "json: expected ',' or ']' at offset " + (offset - 1));
                }
            }
        }

        private String readString() {
            if (atEnd() || text.charAt(offset) != '"') {
                throw new IllegalArgumentException("json: expected a string at offset " + offset);
            }

            offset++;

            StringBuilder out = new StringBuilder();

            while (true) {
                if (atEnd()) {
                    throw new IllegalArgumentException("json: unterminated string");
                }

                char character = text.charAt(offset++);

                if (character == '"') {
                    return out.toString();
                }

                if (character != '\\') {
                    out.append(character);

                    continue;
                }

                if (atEnd()) {
                    throw new IllegalArgumentException("json: truncated escape");
                }

                char escape = text.charAt(offset++);

                switch (escape) {
                    case '"' -> out.append('"');
                    case '\\' -> out.append('\\');
                    case '/' -> out.append('/');
                    case 'b' -> out.append('\b');
                    case 'f' -> out.append('\f');
                    case 'n' -> out.append('\n');
                    case 'r' -> out.append('\r');
                    case 't' -> out.append('\t');
                    case 'u' -> {
                        // Surrogate pairs arrive as two consecutive \\u escapes and
                        // are appended as-is; Java strings are UTF-16, so the pair
                        // reassembles itself.
                        if (offset + 4 > text.length()) {
                            throw new IllegalArgumentException("json: truncated \\u escape");
                        }

                        try {
                            out.append(
                                    (char)
                                            Integer.parseInt(
                                                    text.substring(offset, offset + 4), 16));
                        } catch (NumberFormatException error) {
                            throw new IllegalArgumentException("json: invalid \\u escape", error);
                        }

                        offset += 4;
                    }
                    default ->
                            throw new IllegalArgumentException("json: invalid escape \\" + escape);
                }
            }
        }

        private Double readNumber() {
            int start = offset;

            if (!atEnd() && (text.charAt(offset) == '-' || text.charAt(offset) == '+')) {
                offset++;
            }

            while (!atEnd()) {
                char character = text.charAt(offset);

                if (Character.isDigit(character)
                        || character == '.'
                        || character == 'e'
                        || character == 'E'
                        || character == '+'
                        || character == '-') {
                    offset++;
                } else {
                    break;
                }
            }

            if (start == offset) {
                throw new IllegalArgumentException("json: expected a number at offset " + start);
            }

            return Double.valueOf(text.substring(start, offset));
        }
    }
}

File: sdks/kotlin/src/Json.kt
package dev.lunora

import java.math.BigDecimal
import java.util.Locale

/**
 * A minimal JSON reader/writer plus the stable subscription key.
 *
 * Hand-rolled because the JVM ships no JSON, and pulling in Jackson would make
 * the transport's only dependency a parser it barely uses. The risk that
 * normally makes hand-rolling a bad idea does not apply: the golden fixtures
 * under `protocol/fixtures` are themselves a demanding parser suite, so a
 * mistake here fails the conformance tests rather than hiding.
 *
 * Values map to plain Kotlin types, matching every other Lunora port: `null`,
 * [Boolean], [Double], [String], [List] and [LinkedHashMap] (insertion ordered,
 * because object field order must survive a round trip).
 */
object Json {
    /**
     * Levels of ENVELOPE a wire value can sit under before its own nesting
     * starts. Every payload arrives wrapped: `{"result": V}` is one, and the two
     * deepest — a batch response `{"results":[{"result": V}]}` and a poke part
     * `{"rowsPatch":[{"value": V}]}` — are three.
     */
    private const val MAX_ENVELOPE_DEPTH: Int = 3

    /**
     * The parser's own nesting cap, counted from the DOCUMENT root.
     *
     * It cannot be [Wire.MAX_DEPTH] directly: that bounds a wire VALUE, whose
     * root is already [MAX_ENVELOPE_DEPTH] levels into the frame carrying it, so
     * charging the envelope against the value's budget refused frames whose
     * payload the reference encodes happily. The cap is here to keep a hostile
     * frame off the stack, and a few extra levels cost nothing against that.
     */
    const val MAX_DEPTH: Int = Wire.MAX_DEPTH + MAX_ENVELOPE_DEPTH

    fun parse(text: String): Any? {
        val parser = Parser(text)
        val value = parser.readValue()

        parser.skipWhitespace()

        require(parser.atEnd()) { "json: trailing content at offset ${parser.offset}" }

        return value
    }

    fun write(value: Any?): String = StringBuilder().also { writeValue(it, value) }.toString()

    private fun writeValue(out: StringBuilder, value: Any?) {
        when (value) {
            null -> out.append("null")
            is Boolean -> out.append(if (value) "true" else "false")
            is Number -> out.append(Key.formatNumber(value.toDouble()))
            is String -> out.append(Key.jsonString(value))
            is List<*> -> {
                out.append('[')
                value.forEachIndexed { index, item ->
                    if (index > 0) out.append(',')
                    writeValue(out, item)
                }
                out.append(']')
            }
            is Map<*, *> -> {
                out.append('{')
                value.entries.forEachIndexed { index, entry ->
                    if (index > 0) out.append(',')
                    out.append(Key.jsonString(entry.key.toString()))
                    out.append(':')
                    writeValue(out, entry.value)
                }
                out.append('}')
            }
            else -> throw IllegalArgumentException("json: cannot write a ${value.javaClass.name}")
        }
    }

    private class Parser(private val text: String) {
        var offset: Int = 0
        private var depth: Int = 0

        fun atEnd(): Boolean = offset >= text.length

        fun skipWhitespace() {
            while (offset < text.length && text[offset].isWhitespace()) offset++
        }

        fun readValue(): Any? {
            skipWhitespace()
            require(!atEnd()) { "json: unexpected end of input" }

            // Bounded here, not after parsing. Wire.MAX_DEPTH applies to the
            // decoded tree and never sees the payload; without this a deeply
            // nested frame overflows the stack, and StackOverflowError is an
            // Error that handleFrame's RuntimeException catch cannot catch.
            require(depth <= MAX_DEPTH) { "json: nesting exceeds the $MAX_DEPTH-level limit" }

            return when (text[offset]) {
                '{' -> readObject()
                '[' -> readArray()
                '"' -> readString()
                't' -> readLiteral("true", true)
                'f' -> readLiteral("false", false)
                'n' -> readLiteral("null", null)
                else -> readNumber()
            }
        }

        private fun readLiteral(literal: String, value: Any?): Any? {
            require(text.startsWith(literal, offset)) { "json: invalid literal at offset $offset" }
            offset += literal.length

            return value
        }

        private fun readObject(): Map<String, Any?> {
            val fields = LinkedHashMap<String, Any?>()

            depth++
            offset++ // '{'
            skipWhitespace()

            if (!atEnd() && text[offset] == '}') {
                offset++
                depth--

                return fields
            }

            while (true) {
                skipWhitespace()

                val key = readString()

                skipWhitespace()
                require(!atEnd() && text[offset] == ':') { "json: expected ':' at offset $offset" }
                offset++
                fields[key] = readValue()
                skipWhitespace()
                require(!atEnd()) { "json: unterminated object" }

                when (val character = text[offset++]) {
                    '}' -> {
                        depth--

                        return fields
                    }
                    ',' -> Unit
                    else -> throw IllegalArgumentException("json: expected ',' or '}' but found '$character'")
                }
            }
        }

        private fun readArray(): List<Any?> {
            val items = mutableListOf<Any?>()

            depth++
            offset++ // '['
            skipWhitespace()

            if (!atEnd() && text[offset] == ']') {
                offset++
                depth--

                return items
            }

            while (true) {
                items.add(readValue())
                skipWhitespace()
                require(!atEnd()) { "json: unterminated array" }

                when (val character = text[offset++]) {
                    ']' -> {
                        depth--

                        return items
                    }
                    ',' -> Unit
                    else -> throw IllegalArgumentException("json: expected ',' or ']' but found '$character'")
                }
            }
        }

        private fun readString(): String {
            require(!atEnd() && text[offset] == '"') { "json: expected a string at offset $offset" }
            offset++

            val out = StringBuilder()

            while (true) {
                require(!atEnd()) { "json: unterminated string" }

                val character = text[offset++]

                if (character == '"') return out.toString()

                if (character != '\\') {
                    out.append(character)

                    continue
                }

                require(!atEnd()) { "json: truncated escape" }

                when (val escape = text[offset++]) {
                    '"' -> out.append('"')
                    '\\' -> out.append('\\')
                    '/' -> out.append('/')
                    'b' -> out.append('\b')
                    'f' -> out.append('\u000C')
                    'n' -> out.append('\n')
                    'r' -> out.append('\r')
                    't' -> out.append('\t')
                    'u' -> {
                        // Surrogate pairs arrive as two consecutive \u escapes;
                        // JVM strings are UTF-16, so the pair reassembles itself.
                        require(offset + 4 <= text.length) { "json: truncated \\u escape" }

                        val code = text.substring(offset, offset + 4).toIntOrNull(16)

                        requireNotNull(code) { "json: invalid \\u escape" }
                        out.append(code.toChar())
                        offset += 4
                    }
                    else -> throw IllegalArgumentException("json: invalid escape \\$escape")
                }
            }
        }

        private fun readNumber(): Double {
            val start = offset

            if (!atEnd() && (text[offset] == '-' || text[offset] == '+')) offset++

            while (!atEnd() && (text[offset].isDigit() || text[offset] in ".eE+-")) offset++

            require(start != offset) { "json: expected a number at offset $start" }

            return text.substring(start, offset).toDouble()
        }
    }
}

/**
 * The stable subscription key, ported from `shared/stable-key.ts`.
 *
 * A key is compared verbatim against one produced by the reference TypeScript
 * client, so every spelling must match ECMAScript exactly — a mismatch silently
 * splits one subscription into two.
 */
object Key {
    /** Canonical JSON encoding: keys sorted at every depth, arrays keeping order. */
    fun stableStringify(value: Any?): String = StringBuilder().also { writeStable(it, value) }.toString()

    /** The stable cache/dedup key for [value]. */
    fun stableWireKey(value: WireValue): String = stableStringify(Wire.encode(value))

    private fun writeStable(out: StringBuilder, value: Any?) {
        when (value) {
            null -> out.append("null")
            is Boolean -> out.append(if (value) "true" else "false")
            is Number -> out.append(formatNumber(value.toDouble()))
            is String -> out.append(jsonString(value))
            is List<*> -> {
                out.append('[')
                value.forEachIndexed { index, item ->
                    if (index > 0) out.append(',')
                    writeStable(out, item)
                }
                out.append(']')
            }
            is Map<*, *> -> {
                // The JVM's String.compareTo already compares UTF-16 code units,
                // which is exactly JavaScript's ordering — the only runtime in
                // this set needing no conversion. Go, Ruby, Rust and Swift each
                // required an explicit one.
                val keys = value.keys.map { it.toString() }.sorted()

                out.append('{')
                keys.forEachIndexed { index, key ->
                    if (index > 0) out.append(',')
                    out.append(jsonString(key))
                    out.append(':')
                    writeStable(out, value[key])
                }
                out.append('}')
            }
            else -> out.append("null")
        }
    }

    /**
     * Renders a double exactly as `String(v)` does in JavaScript.
     *
     * Kotlin's `toString` keeps a ".0" on integral values and writes "1.0E-5";
     * ECMAScript drops the decimal, stays positional between 1e-7 and 1e21, and
     * writes "1e-7".
     */
    fun formatNumber(value: Double): String {
        if (value.isNaN() || value.isInfinite()) return "null"

        if (value == Math.rint(value) && Math.abs(value) < 1e21) {
            return BigDecimal.valueOf(value).setScale(0, java.math.RoundingMode.HALF_UP).toBigInteger().toString()
        }

        val magnitude = Math.abs(value)

        if (magnitude >= 1e-6 && magnitude < 1e21) {
            return trimTrailingZeros(BigDecimal(value.toString()).toPlainString())
        }

        return exponential(value)
    }

    private fun exponential(value: Double): String {
        for (precision in 0..17) {
            // Locale.ROOT is load-bearing: the default locale decides the
            // decimal separator, so a German-locale JVM formats 1.5e-7 as
            // "1,5e-07" and produces a different subscription key than the same
            // code elsewhere. The Java port shipped this bug until a test caught it.
            val candidate = String.format(Locale.ROOT, "%.${precision}e", value)

            if (candidate.toDouble() == value) return normaliseExponent(candidate)
        }

        return normaliseExponent(String.format(Locale.ROOT, "%.17e", value))
    }

    /** "1.000000e-07" -> "1e-7": ECMAScript pads neither mantissa nor exponent. */
    private fun normaliseExponent(text: String): String {
        val marker = text.indexOf('e')

        if (marker < 0) return text

        val mantissa = trimTrailingZeros(text.substring(0, marker))
        var exponent = text.substring(marker + 1)
        val sign = if (exponent.startsWith("-")) "-" else "+"

        exponent = exponent.removePrefix("-").removePrefix("+").trimStart('0').ifEmpty { "0" }

        return "${mantissa}e$sign$exponent"
    }

    private fun trimTrailingZeros(text: String): String = if (!text.contains('.')) text else text.trimEnd('0').trimEnd('.')

    /** Quotes a string the way `JSON.stringify` does: `<`, `>`, `&`, U+2028/9 stay raw. */
    fun jsonString(value: String): String {
        val out = StringBuilder(value.length + 2)

        out.append('"')

        for (character in value) {
            when (character) {
                '"' -> out.append("\\\"")
                '\\' -> out.append("\\\\")
                '\n' -> out.append("\\n")
                '\r' -> out.append("\\r")
                '\t' -> out.append("\\t")
                '\b' -> out.append("\\b")
                '\u000C' -> out.append("\\f")
                else ->
                    if (character < ' ') {
                        out.append(String.format(Locale.ROOT, "\\u%04x", character.code))
                    } else {
                        out.append(character)
                    }
            }
        }

        out.append('"')

        return out.toString()
    }
}
